processor/src/operations/sys_ops.rs

```diff
@@ -11,9 +11,132 @@ impl Process {
+        self.stack.shift_left(1);
+
+        Ok(())
+    }
+}
+
+// TESTS
+// ================================================================================================
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        super::{init_stack_with, Operation},
+        Felt, FieldElement, Process,
+    };
+
+    #[test]
+    fn op_fmpupdate() {
+        let mut process = Process::new_dummy();
+
+        // initial value of fmp register should be zero
+        assert_eq!(Felt::ZERO, process.system.fmp());
+
+        // increment fmp register
+        process.execute_op(Operation::Push(Felt::new(2))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(2), process.system.fmp());
+
+        // increment fmp register again
+        process.execute_op(Operation::Push(Felt::new(3))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(5), process.system.fmp());
+
+        // decrement fmp register
+        process.execute_op(Operation::Push(-Felt::new(3))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(2), process.system.fmp());
+
+        // decrementing beyond zero should be an error
+        process.execute_op(Operation::Push(-Felt::new(3))).unwrap();
+        assert!(process.execute_op(Operation::FmpUpdate).is_err());
+
+        // should not affect the rest of the stack state
+        let mut process = Process::new_dummy();
+        init_stack_with(&mut process, &[2, 3]);
+        process.execute_op(Operation::FmpUpdate).unwrap();
+
+        let expected = build_expected(&[2]);
+        assert_eq!(expected, process.stack.trace_state());
+    }
+
+    #[test]
+    fn op_fmpadd() {
+        let mut process = Process::new_dummy();
+
+        // set value of fmp register
+        process.execute_op(Operation::Push(Felt::new(2))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+
+        // compute address of the first local
+        process.execute_op(Operation::Push(-Felt::new(1))).unwrap();
+        process.execute_op(Operation::FmpAdd).unwrap();
+
+        let expected = build_expected(&[1]);
+        assert_eq!(expected, process.stack.trace_state());
+
+        // compute address of second local (also make sure that rest of stack is not affected)
+        process.execute_op(Operation::Push(-Felt::new(2))).unwrap();
+        process.execute_op(Operation::FmpAdd).unwrap();
+
+        let expected = build_expected(&[0, 1]);
+        assert_eq!(expected, process.stack.trace_state());
+    }
+
+    // HELPER FUNCTIONS
+    // --------------------------------------------------------------------------------------------
+
+    fn build_expected(values: &[u64]) -> [Felt; 16] {
+        let mut expected = [Felt::ZERO; 16];
+        for (&value, result) in values.iter().zip(expected.iter_mut()) {
+            *result = Felt::new(value);
+        }
+        expected
+    }
```

Comment:
Not for this PR, but there are a few duplicates of this function floating around (along with a version that takes Felt values instead of u64 that's also duplicated a few times). Is there another way we can handle unit test helpers to avoid this? Or do you feel that it's better to just continue redefining these in the places where we need them? (Is this an idiomatic thing with Rust unit tests?)

Reply:
Yep - there are a few different versions of this, and we should definitely consolidate them. But I'd wait for now to see how many different flavors of this would we need, and then decide on the best way to merge them together.

Reply:
Sounds reasonable to me